src/hls.c

```diff
@@ -0,0 +1,470 @@
+/** @file
+ * @brief HLS output support
+ */
+
+#include <errno.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+#include <libgen.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdlib.h>
+#include <linux/limits.h>
+#include "mumudvb.h"
+#include "errors.h"
+#include "log.h"
+#include "unicast_http.h"
+#include "hls.h"
+#include "ts.h"
+
+static char *log_module="HLS : ";
+
+pthread_mutex_t hls_periodic_lock;
+hls_open_fds_t hls_fds[MAX_CHANNELS];
+extern int dont_send_scrambled;
+
+int mkpath(char* file_path, mode_t mode) {
+    if (!file_path || !*file_path) return -1;
+
+    for (char* p = strchr(file_path + 1, '/'); p; p = strchr(p + 1, '/')) {
+        *p = '\0';
+	log_message( log_module, MSG_FLOOD,"Create dir \"%s\"\n", file_path);
+        if (mkdir(file_path, mode)) {
+            if (errno != EEXIST) {
+                *p = '/';
+                return -1;
+            }
+        }
+        *p = '/';
+    }
+    log_message( log_module, MSG_FLOOD,"Create dir \"%s\"\n", file_path);
+    if (mkdir(file_path, mode)) {
+        if (errno != EEXIST) {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+int hls_sid_entry_compare(const void *pa, const void *pb) {
+	const int *a = pa;
+	const int *b = pb;
+	return a[0] - b[0];
+}
+
+int hls_find_iframe(unsigned char *buf, unsigned int size) {
+	ts_header_t *header;
+	af_header_t *af;
+	unsigned char *h264_packet;
+
+	unsigned int pos = 0;
+
+	// TODO: detection code needs futher testing on various live traffic
+	for (; pos < size; pos += TS_PACKET_SIZE) {
+		header=(ts_header_t *)(buf + pos);
+
+		// MPEG-TS encoder signalling
+		// match 2 (af + no payload) or 3 (af + payload) + RAI = 1
+		af=(af_header_t *)(buf + pos + TS_HEADER_LEN);
+    		if (header->adaptation_field_control > 1 && af->random_access_indicator)
+    		{
+		    log_message( log_module, MSG_FLOOD,"MPEG-TS RAI found at %d\n", pos);
+        	    goto exit;
+    		}
+
+		// h264 stream
+		// match 1 (payload only)
```

Comment:
Is there a way to improve this to also detect MPEG2/h265 i-frame start codes?

Reply:
Can't find understandable documentation on which types of h265 i-frames cutting is allowed.
And in theory, MPEG2 encode must convert i-frames from high-level protocol to RAI by itself. In practice, there is rare case when encoder does not support RAI, but support i-frame.

Reply:
Ah, I see. Makes sense, so normally encoder would just set random access indicator, and the code breaks out at that point. And if that's not available, then manually search for i-frame start based on encoding type.